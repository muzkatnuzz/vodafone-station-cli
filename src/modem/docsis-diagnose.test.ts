import DocsisDiagnose, { BeseitigungBinnenMonatsfrist, downstreamDeviation, DownstreamDeviation64QAM, downstreamDeviationFactory, SofortigeBeseitigung, TolerierteAbweichung, Vorgabekonform } from "./docsis-diagnose";
import type { Diagnose, DocsisStatus, Modulation } from "./modem";
import fixtureDocsisStatus from './__fixtures__/docsisStatus_normalized.json';

test('constructor', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatus as DocsisStatus);
  expect(diagnoser).toBeTruthy();
});

test('checkDownstream', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatus as DocsisStatus);
  const result = diagnoser.checkDownstream() 
  expect(result).toHaveLength(32);
  console.log(result);
  
  expect(result).toMatchSnapshot();
});

test('deviationFactory unsupported modulation', () => {
  expect(()=>downstreamDeviationFactory("8192AM" as Modulation)).toThrowError()
});
test('deviationFactory supported modulation', () => {
  expect(downstreamDeviationFactory("64QAM" as Modulation)).toStrictEqual(new DownstreamDeviation64QAM())
});

test('detectDeviations', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatus as DocsisStatus);
  expect(diagnoser.detectDeviations()).toBeTruthy();
});


describe('diagnoseDownstream', () => {

  test.each`
    input	| expected
    ${{modulation:"64QAM", powerLevel:-60}}	| ${SofortigeBeseitigung}
    ${{modulation:"64QAM", powerLevel:-50}}	| ${SofortigeBeseitigung}
    ${{modulation:"64QAM", powerLevel:-14}}	| ${SofortigeBeseitigung}
    ${{modulation:"64QAM", powerLevel:-13.9}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"64QAM", powerLevel:-13}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"64QAM", powerLevel:-12}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"64QAM", powerLevel:-11.9}}	| ${TolerierteAbweichung}
    ${{modulation:"64QAM", powerLevel:-10.5}}	| ${TolerierteAbweichung}
    ${{modulation:"64QAM", powerLevel:-10}}	| ${TolerierteAbweichung}
    ${{modulation:"64QAM", powerLevel:-9.9}}	| ${Vorgabekonform}
    ${{modulation:"64QAM", powerLevel:-9}}	| ${Vorgabekonform}
    ${{modulation:"64QAM", powerLevel:-8}}	| ${Vorgabekonform}
    ${{modulation:"64QAM", powerLevel:-7}}	| ${Vorgabekonform}
    ${{modulation:"64QAM", powerLevel:0}}	| ${Vorgabekonform}
    ${{modulation:"64QAM", powerLevel:5}}	| ${Vorgabekonform}
    ${{modulation:"64QAM", powerLevel:7}}	| ${Vorgabekonform}
    ${{modulation:"64QAM", powerLevel:7.1}}	| ${TolerierteAbweichung}
    ${{modulation:"64QAM", powerLevel:10}}	| ${TolerierteAbweichung}
    ${{modulation:"64QAM", powerLevel:11.99}}	| ${TolerierteAbweichung}
    ${{modulation:"64QAM", powerLevel:12}}	| ${TolerierteAbweichung}
    ${{modulation:"64QAM", powerLevel:12.1}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"64QAM", powerLevel:13}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"64QAM", powerLevel:14}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"64QAM", powerLevel:14.1}}	| ${SofortigeBeseitigung}
    ${{modulation:"64QAM", powerLevel:100}}	| ${SofortigeBeseitigung}
  `('downstreamDeviation($input)', ({ input, expected }) => {
    expect(downstreamDeviation(input)).toBe(expected);
  });

  test.each`
    input	| expected
    ${{modulation:"256QAM", powerLevel:-60}}	| ${SofortigeBeseitigung}
    ${{modulation:"256QAM", powerLevel:-50}}	| ${SofortigeBeseitigung}
    ${{modulation:"256QAM", powerLevel:-8}}	| ${SofortigeBeseitigung}
    ${{modulation:"256QAM", powerLevel:-7.9}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"256QAM", powerLevel:-7}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"256QAM", powerLevel:-6}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"256QAM", powerLevel:-5.9}}	| ${TolerierteAbweichung}
    ${{modulation:"256QAM", powerLevel:-4.5}}	| ${TolerierteAbweichung}
    ${{modulation:"256QAM", powerLevel:-4}}	| ${TolerierteAbweichung}
    ${{modulation:"256QAM", powerLevel:-3.9}}	| ${Vorgabekonform}
    ${{modulation:"256QAM", powerLevel:-3}}	| ${Vorgabekonform}
    ${{modulation:"256QAM", powerLevel:-2}}	| ${Vorgabekonform}
    ${{modulation:"256QAM", powerLevel:-1}}	| ${Vorgabekonform}
    ${{modulation:"256QAM", powerLevel:0}}	| ${Vorgabekonform}
    ${{modulation:"256QAM", powerLevel:11}}	| ${Vorgabekonform}
    ${{modulation:"256QAM", powerLevel:13}}	| ${Vorgabekonform}
    ${{modulation:"256QAM", powerLevel:13.1}}	| ${TolerierteAbweichung}
    ${{modulation:"256QAM", powerLevel:16}}	| ${TolerierteAbweichung}
    ${{modulation:"256QAM", powerLevel:17.99}}	| ${TolerierteAbweichung}
    ${{modulation:"256QAM", powerLevel:18}}	| ${TolerierteAbweichung}
    ${{modulation:"256QAM", powerLevel:18.1}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"256QAM", powerLevel:19}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"256QAM", powerLevel:20}}	| ${BeseitigungBinnenMonatsfrist}
    ${{modulation:"256QAM", powerLevel:20.1}}	| ${SofortigeBeseitigung}
    ${{modulation:"256QAM", powerLevel:100}}	| ${SofortigeBeseitigung}
  `('downstreamDeviation($input)', ({ input, expected }) => {
    expect(downstreamDeviation(input)).toBe(expected);
  });
});