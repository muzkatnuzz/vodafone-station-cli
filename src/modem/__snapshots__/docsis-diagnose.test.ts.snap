// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkDownstream 1`] = `
Array [
  Object {
    "channelId": "1",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 114,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 55.1,
    "snr": 35,
  },
  Object {
    "channelId": "2",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 130,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 54.7,
    "snr": 35,
  },
  Object {
    "channelId": "3",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 138,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 54.8,
    "snr": 36,
  },
  Object {
    "channelId": "4",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 146,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 54.6,
    "snr": 36,
  },
  Object {
    "channelId": "5",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 602,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 57,
    "snr": 38,
  },
  Object {
    "channelId": "6",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 618,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 57.3,
    "snr": 39,
  },
  Object {
    "channelId": "7",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 626,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 57.7,
    "snr": 38,
  },
  Object {
    "channelId": "8",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 642,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.5,
    "snr": 39,
  },
  Object {
    "channelId": "9",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 650,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.3,
    "snr": 39,
  },
  Object {
    "channelId": "10",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 658,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.3,
    "snr": 39,
  },
  Object {
    "channelId": "11",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 666,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.1,
    "snr": 38,
  },
  Object {
    "channelId": "12",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 674,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.8,
    "snr": 39,
  },
  Object {
    "channelId": "13",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 682,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.8,
    "snr": 39,
  },
  Object {
    "channelId": "14",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 690,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 59.4,
    "snr": 39,
  },
  Object {
    "channelId": "15",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 698,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 53,
    "snr": 35,
  },
  Object {
    "channelId": "16",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 706,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.1,
    "snr": 35,
  },
  Object {
    "channelId": "17",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 714,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.2,
    "snr": 34,
  },
  Object {
    "channelId": "18",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 722,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 53.8,
    "snr": 35,
  },
  Object {
    "channelId": "19",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 730,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 53.9,
    "snr": 35,
  },
  Object {
    "channelId": "20",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 738,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.9,
    "snr": 35,
  },
  Object {
    "channelId": "21",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 746,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 55.3,
    "snr": 35,
  },
  Object {
    "channelId": "22",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 754,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.5,
    "snr": 35,
  },
  Object {
    "channelId": "23",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 762,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.5,
    "snr": 35,
  },
  Object {
    "channelId": "24",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 770,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.5,
    "snr": 35,
  },
  Object {
    "channelId": "25",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 778,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 55,
    "snr": 35,
  },
  Object {
    "channelId": "26",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 786,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 55,
    "snr": 35,
  },
  Object {
    "channelId": "27",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 794,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.9,
    "snr": 34,
  },
  Object {
    "channelId": "28",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 802,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.4,
    "snr": 35,
  },
  Object {
    "channelId": "29",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 810,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.1,
    "snr": 35,
  },
  Object {
    "channelId": "30",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 818,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.5,
    "snr": 35,
  },
  Object {
    "channelId": "31",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 826,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.5,
    "snr": 35,
  },
  Object {
    "channelId": "32",
    "channelType": "SC-QAM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequency": 834,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 53.7,
    "snr": 34,
  },
]
`;

exports[`checkOfdmDownstream 1`] = `
Array [
  Object {
    "channelId": "33",
    "channelType": "OFDM",
    "diagnose": Object {
      "color": "red",
      "description": "SofortigeBeseitigung",
      "deviation": true,
    },
    "frequencyEnd": 324,
    "frequencyStart": 151,
    "lockStatus": "Locked",
    "modulation": "1024QAM",
    "powerLevel": 56.1,
    "snr": 39,
  },
]
`;
